{
 "cells": [
  {
   "cell_type": "raw",
   "id": "732d0270",
   "metadata": {},
   "source": [
    "---\n",
    "title: \"RNA-seq and ATAC Tutorial\"\n",
    "output: \n",
    "  html_document:\n",
    "    toc: true \n",
    "    number_sections: true\n",
    "    toc_depth: 3\n",
    "    toc_float: true \n",
    "    css: scripts/custom.css\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95ed0ae8-9ae5-4992-bd9b-582ec584aa13",
   "metadata": {},
   "outputs": [],
   "source": [
    "filepath <- c(\"/gs/gsfs0/shared-lab/omics-workshop-lab/OmicsWorkshopVignettes/03_BulkRNAATAC_KithXiang/data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "659b66ea-5bd4-47b3-bd71-8c67934434f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########################################################################33\n",
    "#tutorial0_prerequisites\n",
    "######################################################################33\n",
    "\n",
    "#make sure all of these libraries load properly\n",
    "\n",
    "#bioconductor\n",
    "library(\"GenomicRanges\")\n",
    "library(\"DESeq2\")\n",
    "library(\"ACME\")\n",
    "library(\"GEOquery\")\n",
    "library(\"EnhancedVolcano\")\n",
    "\n",
    "#R \n",
    "library(\"stringr\")\n",
    "library(\"dplyr\")\n",
    "library(\"parallel\")\n",
    "library(\"rmarkdown\")\n",
    "library(\"knitr\")\n",
    "library(\"ggfortify\")\n",
    "library(\"data.table\")\n",
    "library(\"ggrepel\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "175c2cf1-60a8-4588-8c79-ec5ef66ff472",
   "metadata": {},
   "outputs": [],
   "source": [
    "#if there were any errors, use this code to install missing libraries\n",
    "#bioconductor packages\n",
    "\n",
    "\n",
    "# bioc.packages = c(\n",
    "#     \"GenomicRanges\",\n",
    "#     \"DESeq2\",\n",
    "#     \"ACME\",\n",
    "#     \"GEOquery\",\n",
    "#     \"EnhancedVolcano\"\n",
    "# )\n",
    "# if (!require(\"BiocManager\", quietly = TRUE))\n",
    "#     install.packages(\"BiocManager\")\n",
    "# BiocManager::install(bioc.packages)\n",
    "\n",
    "\n",
    "# #R packages\n",
    "# r.packages = c(\n",
    "#     \"stringr\",\n",
    "#     \"dplyr\",\n",
    "#     \"parallel\",\n",
    "#     \"rmarkdown\",\n",
    "#     \"knitr\",\n",
    "#     \"ggfortify\",\n",
    "#     \"data.table\",\n",
    "#     \"ggrepel\"\n",
    "# )\n",
    "# ix.installed = r.packages %in% installed.packages()[,1]\n",
    "# install.packages(r.packages[!ix.installed])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1ccc7c7-73a2-4888-9ec8-4b7dd42d5c60",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################\n",
    "#tutorial1_rna\n",
    "#########################################33333\n",
    "\n",
    "#########\n",
    "#RNA section 2.6\n",
    "\n",
    "#get the count files\n",
    "pfiles = list.files(path = paste0(filepath,\"/counts_rna\"),\n",
    "    pat = \".*star.*.tab\", full = TRUE, recursive = TRUE)\n",
    "\n",
    "#load the count files and just use the first 2 columns\n",
    "pcounts = lapply(pfiles, function(f){\n",
    "    print(f)\n",
    "    try({\n",
    "        x = read.csv(f, sep=\"\\t\", header=F)\n",
    "        #remove bad rows\n",
    "        x = x[grepl(x[,1], pat=\"ENSG\"),]\n",
    "        y = data.frame(ensg=x[,1], counts=x[,2])\n",
    "        y\n",
    "    }, silent=T)\n",
    "})\n",
    "\n",
    "#remove NAs\n",
    "names(pcounts) = pfiles\n",
    "ix = unlist(lapply(pcounts, function(a){class(a) != \"try-error\"}))\n",
    "pcounts = pcounts[ix]\n",
    "is.na(pcounts)\n",
    "\n",
    "#merge into a data.frame\n",
    "x = do.call(cbind, lapply(pcounts, function(a){\n",
    "\ta$counts\n",
    "}))\n",
    "#make the names easier to read\n",
    "colnames(x) = gsub(names(pcounts), pat=\".*/(.*)_ReadsPerGene.out.tab\", rep=\"\\\\1\")\n",
    "rownames(x) = pcounts[[1]][,1]\n",
    "\n",
    "\n",
    "#get rid of NA counts\n",
    "ix2 = which(apply(x, 1, function(a){\n",
    "    !any(is.na(a))\n",
    "}))\n",
    "x=x[ix2,]\n",
    "\n",
    "#take out the rownames that have __ in them\n",
    "#only use the ENSG rows\n",
    "x = x[grepl(rownames(x), pat=\"ENSG\"),]\n",
    "\n",
    "#take out the dot?\n",
    "rownames(x) = gsub(rownames(x), pat=\"\\\\..*\", rep=\"\")\n",
    "\n",
    "head(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b44f860b-8d8c-43fc-827b-767fc7f633e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "#RNA section 3.1\n",
    "\n",
    "# #create a folder/directory for where geo downloads will be cached\n",
    "# if (!dir.exists(\"geo\")){\n",
    "#     dir.create(\"geo\")\n",
    "# }\n",
    "\n",
    "# #download geo data with id GSE132040 and save to \"geo\" dir\n",
    "# dat <- getGEO(\"GSE132040\", destdir = \"geo\")\n",
    "# head(dat)\n",
    "\n",
    "# use local GEO files \n",
    "dat <- vector(\"list\",length = 1)\n",
    "names(dat) <- \"GSE132040_series_matrix.txt.gz\"   \n",
    "dat[[1]] <-getGEO(filename= paste0(filepath,\"/geo/GSE132040_series_matrix.txt.gz\"))\n",
    "head(dat)\n",
    "\n",
    "#load the phenotype data from GEO, obtain phenodata\n",
    "#some GSE's have more than 1 of dataset, getGEO returns them in a list\n",
    "phenodata <- pData(dat[[1]])\n",
    "head(phenodata)\n",
    "#Alternatively, you can do the following \n",
    "#phenodata <- dat[[\"GSE132040_series_matrix.txt.gz\"]]@phenoData@data\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d614755d-2cce-4d19-9f74-172602319f9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############\n",
    "#RNA section 3.2\n",
    "\n",
    "#Load the expression data\n",
    "rnadata.geo <- exprs(dat[[1]])\n",
    "dim(rnadata.geo) #0 rows?!\n",
    "\n",
    "#the columns of rnadata should match up to the rows of phenodata)\n",
    "#cbind(colnames(rnadata),rownames(phenodata))\n",
    "all(colnames(rnadata.geo) == rownames(phenodata))\n",
    "\n",
    "\n",
    "\n",
    "#if you want to use read.csv you'll need to unzip the file first\n",
    "#rnadata <- read.csv(\"GSE132040_190214_A00111_0269_AHH3J3DSXX_190214_A00111_0270_BHHMFWDSXX.csv\")\n",
    "\n",
    "#here's a trick to load large files using the data.table function.  It's much faster than read.csv and even works on zipped files\n",
    "rnadata <- data.frame(fread(paste0(filepath,\"/GSE132040_190214_A00111_0269_AHH3J3DSXX_190214_A00111_0270_BHHMFWDSXX.csv.gz\")))\n",
    "dim(rnadata)\n",
    "\n",
    "#look at the first 5 rows and first 10 columns to get an idea of the data\n",
    "#View(rnadata[1:5, 1:10])\n",
    "#print(rnadata[1:5, 1:10])\n",
    "rnadata[1:5, 1:10]\n",
    "\n",
    "\n",
    "table(duplicated(rnadata$gene)) # check if any duplication gene names \n",
    "\n",
    "\n",
    "rownames(rnadata) <- rnadata$gene  # Assign the first column to the rownames\n",
    "rnadata$gene <- NULL #remove the column from the rest of the matrix\n",
    "\n",
    "class(rnadata)\n",
    "\n",
    "rnadata[1:5, 1:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab274606-d7a4-4a54-b75d-7a49bafd234d",
   "metadata": {},
   "outputs": [],
   "source": [
    "################\n",
    "#RNA section 3.3\n",
    "\n",
    "cbind(head(colnames(rnadata)))\n",
    "\n",
    "cbind(head(phenodata$title))\n",
    "\n",
    "# we can use gsub to manipulate the string. in this case, replace a part of a string.\n",
    "colnames(rnadata) <- gsub(colnames(rnadata), pattern=\".gencode.vM19\", replace=\"\")\n",
    "                     \n",
    "cbind(head(colnames(rnadata))) # after\n",
    "\n",
    "\n",
    "# create a new column in phenodata called \"tmp_id\" \n",
    "# that has the sample identifier extracted from the title\n",
    "# we can use replace again noticed I used \\\\ in front of ( ) and [ ], because they are special characters \n",
    "# a more general approach is to use regular expression \n",
    "phenodata$tmp_id <- gsub(phenodata$title, pattern=\"Tabula Muris Senis \\\\(bulk RNA-seq\\\\) \\\\[\", replace=\"\")\n",
    "phenodata$tmp_id <- gsub(phenodata$tmp_id, pattern=\"\\\\]\", replace=\"\")\n",
    "\n",
    "cbind(head(phenodata$tmp_id))\n",
    "\n",
    "# lets also get rid of :chr1 in some colnames\n",
    "colnames(phenodata) <- gsub(colnames(phenodata), pattern=\":ch1\", replace=\"\")\n",
    "\n",
    "\n",
    "#is every sample accounted for? yes\n",
    "all(phenodata$tmp_id %in% colnames(rnadata))\n",
    "\n",
    "#but they don't match\n",
    "head(cbind(phenodata$tmp_id, colnames(rnadata)))\n",
    "\n",
    "#we need to reorder the rows of phenodata to match the columns of rnadata\n",
    "\n",
    "#better to work with a temporary variable so you can check your work\n",
    "phenodata_reordered <- phenodata[match(colnames(rnadata), phenodata$tmp_id),]\n",
    "\n",
    "#check it again (now they match)\n",
    "head(cbind(phenodata_reordered$tmp_id, colnames(rnadata)))\n",
    "\n",
    "#from now on we will replace phenodata with phenodata_reordered\n",
    "phenodata <- phenodata_reordered\n",
    "rm(phenodata_reordered)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce895865-ac10-4e0e-940e-1d2f1bf0529c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################\n",
    "#RNA section 3.4\n",
    "\n",
    "#run deseq on a subset of rnadata for demonstration purposes\n",
    "table(phenodata$tissue, phenodata$age)\n",
    "\n",
    "#lets just look at bone vs brain\n",
    "#in 1 months postnatal mice\n",
    "\n",
    "#extract the relevant rows from phenotype data\n",
    "phenodata_small <- phenodata[phenodata$tissue %in% c(\"Bone\", \"Brain\"),]\n",
    "phenodata_small <- phenodata_small[phenodata_small$age %in% c(\"1 months postnatal\"),]\n",
    "\n",
    "\n",
    "#extract the relevant columns from count data\n",
    "rnadata_small <- rnadata[,colnames(rnadata) %in% phenodata_small$tmp_id]\n",
    "\n",
    "dim(phenodata_small)\n",
    "dim(rnadata_small)\n",
    "\n",
    "#make sure the IDs match up\n",
    "table(phenodata_small$tmp_id == colnames(rnadata_small))\n",
    "\n",
    "#use the new ids instead of the GSM ids\n",
    "#deseq requires the rownames from the info match the colnames of the counts\n",
    "rownames(phenodata_small) <- phenodata_small$tmp_id\n",
    "\n",
    "\n",
    "#make sure the column names don't have any special characters like \":\"\n",
    "colnames(phenodata_small) <- make.names(colnames(phenodata_small))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efda496d-a4ae-4424-8113-a81fc9c9316c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################\n",
    "#RNA section 4.1\n",
    "\n",
    "#it's always a good idea to explicitly set a categorical variable as a factor\n",
    "#so you can control which is the baseline reference.  Otherwise, it sorts \n",
    "#alphabetically and chooses the first as the ref.\n",
    "phenodata_small$tissue <- factor(phenodata_small$tissue, levels = c(\"Bone\", \"Brain\"))\n",
    "\n",
    "dds <- DESeqDataSetFromMatrix(countData=rnadata_small, colDat=phenodata_small, design = ~tissue)\n",
    "dds <- DESeq(dds) #takes a minute\n",
    "res <- results(dds)\n",
    "res <- as.data.frame(res) #easier to work with data.frames\n",
    "\n",
    "#take a look at the results in excel\n",
    "write.table(res, file=\"res.csv\",sep = \",\",quote = F,row.names = T, col.names = NA)\n",
    "\n",
    "summary(res)\n",
    "\n",
    "head(res, 30)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841ea58d-734a-4237-8ecf-2a9c1f4ca8f7",
   "metadata": {},
   "outputs": [],
   "source": [
    "###############\n",
    "#RNA section 4.2\n",
    "\n",
    "#remove results with no counts\n",
    "res <- res[res$baseMean > 0,]\n",
    "#sort by fold change and \n",
    "res <- res[order(res$log2FoldChange),] # turn decreasing = TRUE if want from the largest to the smallest \n",
    "#write the output to a spreadsheet\n",
    "\n",
    "write.table(res,file=\"de_boneVsbrain_age1month.csv\",sep = \",\",quote = F,row.names = T,col.names = NA)\n",
    "\n",
    "dim(res)\n",
    "\n",
    "#how many genes show significant DE after adjusting for multiple comparisons?\n",
    "sum(res$padj < .01, na.rm=T)\n",
    "\n",
    "#what are the genes that have  a log2FC > 2 and an adjusted pvalue < 0.01?\n",
    "res_up <- res[!is.na(res$log2FoldChange),]\n",
    "res_up <- res_up[res_up$log2FoldChange > 2,]\n",
    "res_up <- res_up[!is.na(res_up$padj),]\n",
    "res_up <- res_up[res_up$padj < 0.01,]\n",
    "res_up <- res_up[order(res_up$log2FoldChange,decreasing = TRUE),]\n",
    "\n",
    "res_down <- res[!is.na(res$log2FoldChange),]\n",
    "res_down <- res_down[res_down$log2FoldChange < -2,]\n",
    "res_down <- res_down[!is.na(res_down$padj),]\n",
    "res_down <- res_down[res_down$padj < 0.01,]\n",
    "\n",
    "\n",
    "nrow(res_up)\n",
    "nrow(res_down)\n",
    "\n",
    "#look at the top 20 in each\n",
    "head(rownames(res_up), 20)\n",
    "head(rownames(res_down), 20)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81d87f8c-d57f-45ee-a0cc-408c2d9e997d",
   "metadata": {},
   "outputs": [],
   "source": [
    "##################\n",
    "#RNA section 4.3\n",
    "\n",
    "# when the expression data is alinged to the phenotype data\n",
    "# it's easy to run all the usual informatics procedures \n",
    "dim(rnadata_small)\n",
    "\n",
    "\n",
    "# the prcomp function complains when there are columns with 0 variance\n",
    "# to fix it, remove the genes that have no variance\n",
    "gene_var <- apply(rnadata_small, 1, var) > 0 \n",
    "table(gene_var)\n",
    "\n",
    "rnadata_filtered <- rnadata_small[gene_var,]\n",
    "dim(rnadata_filtered)\n",
    "pca1 <- prcomp(t(rnadata_filtered), scale=T)\n",
    "\n",
    "\n",
    "#plot the first 2 principal components, colored by tissue type\n",
    "#pdf(\"test1.pdf\")\n",
    "autoplot(pca1, data=phenodata_small, colour='tissue', shape=\"Sex\")\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "315a4987-4be7-4e9b-9139-37fd1d113a20",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################\n",
    "#RNA section 4.4\n",
    "\n",
    "vlnPlot1 <- EnhancedVolcano(res,\n",
    "    lab = rownames(res),\n",
    "    #subtitle = NULL, #get rid of the subtitle\n",
    "    #colCustom = keyvals.colour, #give customized color\n",
    "    x = 'log2FoldChange',\n",
    "    y = 'padj',\n",
    "    #xlim = c(-30,30), #x axis range\n",
    "    #ylim = c(0,300), # y axis range\n",
    "    title = 'DE_Bone_vs_Brain: 1month', # title label\n",
    "    xlab= bquote(~Log[2]~ 'fold change'), # x axis label\n",
    "    ylab= bquote(~-Log[10]~ 'Padj'), # y axis label\n",
    "    caption = NULL,\n",
    "    pCutoff = 0.05,\n",
    "    FCcutoff = 2,\n",
    "    pointSize = 2.0)\n",
    "\n",
    "#pdf(\"test2.pdf\")\n",
    "vlnPlot1\n",
    "#dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dd687a2-4068-4f67-8827-bf33b6c6c8eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################################\n",
    "#tutorial2_atac\n",
    "#########################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "509a41f8-1409-40f7-bb1d-d87697ebbf08",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############\n",
    "#ATAC section 2.6.1\n",
    "\n",
    "#load in the peak files\n",
    "pfiles = list.files(path=paste0(filepath,\"/macsResults\"), recursive=T, pat=\"^.*xls$\", full=T)\n",
    "peaks = lapply(pfiles, read.table, header=T)\n",
    "\n",
    "#contruct genomic ranges of the peak summits +/- 200bp\n",
    "ranges = lapply(peaks, function(peak){\n",
    "\tGRanges(seqnames = peak[,1],\n",
    "        strand=rep(\"*\", nrow(peak)),\n",
    "            IRanges(start = peak[,2], end = peak[,3]))\n",
    "})\n",
    "peaks.reduced = reduce(do.call(c, ranges))\n",
    "\n",
    "#save genomic range as a bed file\n",
    "#https://www.biostars.org/p/89341/\n",
    "grange2bed <- function(gr, bfile=\"foo.bed\"){\n",
    "\n",
    "\tdf <- data.frame(seqnames=seqnames(gr),\n",
    "\t\tstarts=start(gr)-1,\n",
    "\t\tends=end(gr),\n",
    "\t\tnames=c(rep(\".\", length(gr))),\n",
    "\t\tscores=c(rep(\".\", length(gr))),\n",
    "\t\tstrands=strand(gr))\n",
    "\n",
    "\twrite.table(df, file=bfile, quote=F, sep=\"\\t\", row.names=F, col.names=F)\n",
    "}\n",
    "\n",
    "grange2bed(peaks.reduced, bfile=\"peaks_stat.bed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0c3ff64-ba7d-4446-a220-d03ff0528b4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "##############\n",
    "#ATAC section 2.7\n",
    "pfiles = list.files(path=paste0(filepath,\"/counts_atac\"), pat=\"*counts\", full=T)\n",
    "\n",
    "#load in 1 file to have a look at the format\n",
    "x = read.csv(pfiles[1], sep=\"\\t\", header=F)\n",
    "head(x)\n",
    "\n",
    "#process all the count files\n",
    "pcounts = lapply(pfiles, function(file){\n",
    "    #print(file)\n",
    "\tx = read.csv(file, sep=\"\\t\", header=F)\n",
    "    #extract the number that appears after the *| in the 6th column of the input\n",
    "\thits = as.numeric(gsub(x[,6], pat=\"\\\\*\\\\|\", rep=\"\"))\n",
    "\ty = data.frame(chr=x[,1], start=x[,2], end=x[,3], counts=hits)\n",
    "\ty\n",
    "})\n",
    "#bind the counts together as columns in a matrix\n",
    "peakdata = do.call(cbind, lapply(pcounts, function(a){\n",
    "\ta$counts\n",
    "}))\n",
    "#give the columns better looking names. \n",
    "#grab the strings between the last \"/\" and \".counts\" \n",
    "colnames(peakdata) = gsub(pfiles, pat=\".*/(.*).counts\", rep=\"\\\\1\")\n",
    "\n",
    "\n",
    "#peak locations\n",
    "peaks = pcounts[[1]][,1:3]\n",
    "head(peaks)\n",
    "#make a descriptive string ID for the peak \n",
    "peaknames = sapply(1:nrow(peaks), function(i){\n",
    "\tpaste0(peaks[i,1:3], collapse=\"_\")\n",
    "})\n",
    "rownames(peakdata) = peaknames\n",
    "cbind(head(rownames(peakdata)))\n",
    "\n",
    "\n",
    "\n",
    "#get rid of NA counts\n",
    "ix2 = which(apply(peakdata, 1, function(a){\n",
    "    !any(is.na(a))\n",
    "}))\n",
    "peakdata=peakdata[ix2,]\n",
    "\n",
    "dim(peakdata)\n",
    "head(peakdata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9573670f-6053-47bd-a498-34f4493217f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "#ATAC section 3\n",
    "\n",
    "info <- colnames(peakdata)\n",
    "comp1 <- rep(NA, length(info))\n",
    "comp1[1:2] <- \"ctrl\"\n",
    "comp1[3:4] <- \"drug_treated\"\n",
    "coldata <- data.frame(info, comp1)\n",
    "\n",
    "#always make sanity checks with your own data\n",
    "#the row order of coldata needs to match the column order of peakdata\n",
    "all(coldata$info == colnames(peakdata))\n",
    "head(coldata)\n",
    "\n",
    "\n",
    "#run the analysis\n",
    "dds <- DESeqDataSetFromMatrix(countData = peakdata,\n",
    "\t\t\t\t\t\t\t\t  colData = coldata,\n",
    "\t\t\t\t\t\t\t\t  design = ~ comp1)\n",
    "dds <- DESeq(dds)\n",
    "res <- results(dds)\n",
    "res <- as.data.frame(res)\n",
    "\n",
    "head(res, 50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "971a5910-25f3-4f02-8a16-eb807716b5a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "#ATAC section 3.1\n",
    "\n",
    "\n",
    "#run DESeq, add peak info, sort and write as a csv.\n",
    "#countdata: a matrix of integer counts, (nPeak x nSamples)\n",
    "#coldata: a data.frame with info on the samples (nSamples x nVariables)\n",
    "#ix: the indices of the samples you want to use in the analysis.\n",
    "#formula: shows which columns of coldata you want to use in the analysis\n",
    "#resFile: is where to save the result file\n",
    "runDE <- function(countdata, coldata, ix, formula = formula(\"~ ep300\"), resFile){\n",
    "\n",
    "\tdds = DESeqDataSetFromMatrix(countData = countdata[,ix],\n",
    "\t\t\t\t\t\t\t\t  colData = coldata[ix,],\n",
    "\t\t\t\t\t\t\t\t  design = formula)\n",
    "\tdds <- updateObject(dds)\n",
    "\tdds = DESeq(dds)\n",
    "\tres = results(dds)\n",
    "\ty = data.frame(res)\n",
    "\n",
    "    #break the peak identifier into parts\n",
    "    pos = data.frame(pid = rownames(y), str_split_fixed(rownames(y), \"_\", 3))\n",
    "    colnames(pos)[2:4] = c(\"chrom\", \"start\", \"stop\")\n",
    "    pos$start = as.numeric(pos$start)\n",
    "    pos$stop = as.numeric(pos$stop)\n",
    "    pos$mid = round((pos$stop+pos$start)/2)\n",
    "\ty = cbind(pos, y)\n",
    "\n",
    "    #add peak counts\n",
    "\tdiffexp = merge(y, countdata[,ix], by.x=1, by.y=0, all=T)\n",
    "\n",
    "\t#append the normalized counts to the matrix\n",
    "\tnorm.counts = counts(dds, normalized=T)\n",
    "\tcolnames(norm.counts) = paste0(\"norm_\",dds$info)\n",
    "\ty = merge(diffexp, norm.counts, by.x=1, by.y=0, all=T)\n",
    "\n",
    "    #sort results by fold change\n",
    "\ty = y[order(y[,\"log2FoldChange\"]),]\n",
    "\n",
    "    #add a chromosomal position to make IGV navigation easier\n",
    "    y$chrpos = paste0(y$chrom, \":\", y$start, \"-\", y$stop)\n",
    "\n",
    "\twrite.table(file=resFile, x=y, row.names=F, sep=\"\\t\")\n",
    "\n",
    "    #also return the results\n",
    "    y\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c072c630-a511-417d-bf14-a2c820fb9718",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "#ATAC section 3.2\n",
    "\n",
    "#use the non-NA samples in comp1\n",
    "ix = which(!is.na(coldata$comp1))\n",
    "#or you can do it like\n",
    "#ix = 1:4\n",
    "\n",
    "#run the analysis\n",
    "res = runDE(peakdata, coldata, ix, formula(\"~ comp1\"), \"de_CtrlVsDrug.txt\")\n",
    "\n",
    "#view some of the results\n",
    "head(res, 50)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0f9ac328-0c2f-4483-9b35-1071bbe20c35",
   "metadata": {},
   "outputs": [],
   "source": [
    "##########\n",
    "#ATAC section 3.3\n",
    "\n",
    "appendClosestGeneInfo <- function(x, chr=1, pos = 2, genome=\"hg19\", numCores = 1){\n",
    "\tres = mclapply(1:nrow(x), function(i){\n",
    "\t\tif (i %% 1000 == 0){\n",
    "\t\t\tprint(i)\n",
    "\t\t}\n",
    "\t\tgenes = findClosestGene(x[i,chr],x[i, pos],genome)\n",
    "\t\t#if there's more than one, just take the first\n",
    "\t\tdata.frame(genes[1,])\n",
    "\t}, mc.cores = numCores)\n",
    "\n",
    "\tx.res = bind_rows(res)\n",
    "\tcbind(x.res, x)\n",
    "}\n",
    "\n",
    "#how many peaks show significant DE?\n",
    "ix.sig = which(res$padj < 0.05)\n",
    "length(ix.sig)\n",
    "\n",
    "#show off the function using a small portion of the data\n",
    "res2 = appendClosestGeneInfo(x=res[ix.sig,], chr=2, pos=5, genome=\"hg38\") \n",
    "\n",
    "#if you get an error, try running on a single thread\n",
    "#res2 = appendClosestGeneInfo(x=res[ix.sig,], chr=2, pos=5, genome=\"hg38\", numCores=1) \n",
    "\n",
    "#if you still get an error, there may be issues with the web host(see below)\n",
    "\n",
    "head(res2, 50)\n",
    "write.csv(file=\"deAnno_padj0.05_CtrlVsDrug.txt\", x=res2, row.names=F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "80dda3e5-5b30-422d-938e-1f31a1173f1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#################################\n",
    "#I've run into intermittent problems with the code above.\n",
    "#the original site that hosts the reference database occasionally goes offline\n",
    "#If you still want to use the code above, you can hack some of the functions inside\n",
    "#the ACME package to make it point to a different host\n",
    "\n",
    "\n",
    "# mygetRefflat <- function (genome = \"hg38\"){\n",
    "#     tmpfile <- tempfile()\n",
    "#     download.file(paste(\"http://hgdownload.soe.ucsc.edu/goldenPath/\", \n",
    "#         genome, \"/database/refFlat.txt.gz\", sep = \"\"), tmpfile, \n",
    "#         mode = \"wb\")\n",
    "#     rf <- read.delim(conn <- gzfile(tmpfile, open = \"rt\"), header = FALSE, \n",
    "#         sep = \"\\t\")\n",
    "#     close(conn)\n",
    "#     colnames(rf) <- c(\"geneName\", \"name\", \"chrom\", \"strand\", \n",
    "#         \"txStart\", \"txEnd\", \"cdsStart\", \"cdsEnd\", \"exonCount\", \n",
    "#         \"exonStarts\", \"exonEnds\")\n",
    "#     txEndNeg <- rf$txStart\n",
    "#     txStartNeg <- rf$txEnd\n",
    "#     cdsStartNeg <- rf$cdsEnd\n",
    "#     cdsEndNeg <- rf$cdsStart\n",
    "#     NegStrand <- rf$strand == \"-\"\n",
    "#     rf[NegStrand, \"cdsEnd\"] <- cdsEndNeg[NegStrand]\n",
    "#     rf[NegStrand, \"cdsStart\"] <- cdsStartNeg[NegStrand]\n",
    "#     rf[NegStrand, \"txEnd\"] <- txEndNeg[NegStrand]\n",
    "#     rf[NegStrand, \"txStart\"] <- txStartNeg[NegStrand]\n",
    "#     return(rf)\n",
    "# }\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5572977-c973-4aa2-8f6b-3e8aa3fa91bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# myfindClosestGene <- function (chrom, pos, genome = \"hg38\", position = \"txStart\"){\n",
    "#     if (!exists(\"refflat\")) {\n",
    "#         reftmp <- list()\n",
    "#         reftmp[[genome]] <- mygetRefflat(genome)\n",
    "#         assign(\"refflat\", reftmp, .GlobalEnv)\n",
    "#     }\n",
    "#     else if (!(genome %in% names(refflat))) {\n",
    "#         refflat[[genome]] <<- mygetRefflat(genome)\n",
    "#     }\n",
    "#     rf <- refflat[[genome]]\n",
    "#     chromsub <- rf$chrom == chrom\n",
    "#     diffdist <- rf[chromsub, position] - pos\n",
    "#     sub <- which(abs(diffdist) == min(abs(diffdist)))\n",
    "#     rf <- rf[chromsub, 1:9][sub, ]\n",
    "#     return(data.frame(rf, Distance = diffdist[sub]))\n",
    "# }\n",
    "\n",
    "# #now you can try again\n",
    "# appendClosestGeneInfo2 <- function(x, chr=1, pos = 2, genome=\"hg38\", numCores = 4){\n",
    "# \tres = mclapply(1:nrow(x), function(i){\n",
    "# \t\tif (i %% 1000 == 0){\n",
    "# \t\t\tprint(i)\n",
    "# \t\t}\n",
    "# \t\tgenes = myfindClosestGene(x[i,chr],x[i, pos],genome)\n",
    "# \t\t#if there's more than one, just take the first\n",
    "# \t\tdata.frame(genes[1,])\n",
    "# \t}, mc.cores = numCores)\n",
    "\n",
    "# \tx.res = bind_rows(res)\n",
    "# \tcbind(x.res, x)\n",
    "# }\n",
    "\n",
    "\n",
    "# res2 = appendClosestGeneInfo2(x=res[ix.sig,], chr=2, pos=5, genome=\"hg38\") \n",
    "# head(res2, 50)\n",
    "# write.csv(file=\"deAnno_padj0.05_CtrlVsDrug.txt\", x=res2, row.names=F)\n"
   ]
  }
 ],
 "metadata": {
  "jupytext": {
   "cell_metadata_filter": "fig.height,file,cache,tags,name,message,fig.width,eval,language,-all",
   "main_language": "R",
   "notebook_metadata_filter": "-all"
  },
  "kernelspec": {
   "display_name": "R [conda env:omics-workshop-R]",
   "language": "R",
   "name": "conda-env-omics-workshop-R-r"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
